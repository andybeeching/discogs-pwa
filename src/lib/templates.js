import rbl from 'remove-blank-lines'
import offlineImg from '../img/relax.gif'
import fourOhFourImg from '../img/404.gif'
import fiveOhOhImg from '../img/500.gif'

// removes superflous line breaks and leading/trailing whitespace
const trim = str => rbl(str).trim()

/**
 * Render the Root page
 *
 * @param {Object} data - Discogs Search Result data entity
 * @returns {String}
 */
export function root(data) {
  return trim(`
    <h2>Popular on Discogs</h2>
    <ul class="grid">
      ${data.results.map(item => getGridItem(item, item.type)).join('')}
    </ul>
  `)
}

/**
 * Render the Search Results page content
 *
 * @param {Object} data - Discogs Search Result data entity
 * @param {String} query - Search terms (unencoded)
 * @returns {String}
 */
export function artistsSearchResults(data, query) {
  const resultRange = getResultRange(data.pagination)

  const slug = `/search/${encodeURIComponent(query)}/`
  const pagination = getPagination(data.pagination, slug)

  return trim(`
    ${query ? `<h2>Search results for "${query}"</h2>` : ''}
    ${resultRange}
    ${pagination}
    <ul class="grid">
      ${data.results
        .map(item => getGridItem(item, 'artist', '/page/1'))
        .join('')}
    </ul>
    ${pagination}
  `)
}

/**
 * Render the number of search results with range
 *
 * @param {Object} pagination - Discogs Pagination data entity
 * @docs https://www.discogs.com/developers#page:home,header:home-pagination
 * @returns {String}
 */
export function getResultRange(pagination) {
  const { items, page, pages, per_page: perPage } = pagination
  const startRange = (page - 1) * perPage + 1
  const endRange = page === pages ? items : page * perPage

  return `
    ${
      items > 0
        ? `
      <p>${pagination.items} results were found, viewing ${startRange} to ${endRange}</p>
    `
        : 'No search results found'
    }
  `
}

/**
 *
 * @param {Object} pagination - Discogs Pagination data entity
 * @docs https://www.discogs.com/developers#page:database,header:database-search-get
 * @param {String} slug - URL encoded PWA endpoint for related requests
 * @example 'search?q=The%20Beatles'
 * @returns {String}
 */
export function getPagination(pagination, slug) {
  const { page, pages } = pagination

  return pages > 1
    ? `
    <ol class="pagination">
      <li>
        ${
          page === 1
            ? `<span class="disabled-link">&laquo;</span>`
            : `<a href="${slug}page/1">&laquo;</a>`
        }
      </li>
      <li>
        ${
          page === 1
            ? `<span class="disabled-link">Previous page</span>`
            : `<a href="${slug}page/${pagination.page - 1}">Previous page</a>`
        }
      </li>
      <li>
        ${
          page === pages
            ? `<span class="disabled-link">Next page</span>`
            : `<a href="${slug}page/${pagination.page + 1}">Next page</a>`
        }
      </li>
      <li>
        ${
          page === pages
            ? `<span class="disabled-link">&raquo;</span>`
            : `<a href="${slug}page/${pages}">&raquo;</a>`
        }
      </li>
    </ol>
  `
    : ''
}

/**
 * Renders Grid item
 *
 * @param {Object} item - Discogs Artist or Release data entity
 * @param {String} slug - URL encoded PWA endpoint for related requests
 * @example 'artist'
 * @param {String} trailing - additional string to append to releated URLs
 * @optional
 * @example 'page=1'
 * @returns {String}
 */
function getGridItem(item, slug, trailing = '') {
  return `
    <li class="grid-item">
      <a href="/${slug}/${item.id}${trailing}">
        <img src="${item.thumb}" />
        <span class="desc">${item.title}</span>
      </a>
    </li>
  `
}

/**
 * Renders a list of artists as a readable list
 *
 * @param {Array} artists - array of 'members' from Artist data entity
 * @example [{ name: {String}, id: {Number}}, ...]
 * @returns {String}
 */
export function explodeArtists(artists) {
  return artists
    .map((item, idx, arr) => {
      const isLast = arr[idx + 1] === undefined
      const leadingChars = idx > 0 && isLast ? 'and ' : ''
      const trailingChars = !isLast && !leadingChars ? ', ' : ''

      return `${leadingChars}<a href="/artist/${item.id}/page/1">${item.name}</a>${trailingChars}`
    })
    .join('')
}

/**
 * Renders an artist summary
 *
 * @param {{Object}} data - Discogs Artist data entity
 * @returns {String}
 */
export function artist(data) {
  return `
    <h2 class="h2">${data.name}</h2>
    <p>${data.profile}</p>
    ${data.members ? `<p>Members: ${explodeArtists(data.members)}</p>` : ''}
  `
}

/**
 * Renders artist releases in descending (by release date) order.
 *
 * @param {Object} data - Discogs Artist Releases data entity
 * @param {Number} artistId - artist id
 * @returns {String}
 */
export function releaseList(data, artistId) {
  const slug = `/artist/${artistId}/`
  const pagination = getPagination(data.pagination, slug)

  return trim(`
    <h3 class="h3">Releases</h2>
    ${pagination}
    <ul class="grid">
      ${data.releases.map(item => getGridItem(item, item.type)).join('')}
    </ul>
    ${pagination}
  `)
}

/**
 * Renders the release page
 *
 * @param {Object} data - Discogs Release data entity
 * @returns {String}
 */
export function release(data) {
  const { artists, tracklist, title, images } = data

  return trim(`
    <h2 class="h2">${title}</h2>
    <p>By ${explodeArtists(artists)}</p>
    <div>
      <img src="${images[0].resource_url}" class="artwork" />
    </div>
    <h3>Tracklist</h3>
    <ol>
        ${tracklist.map(item => `<li>${item.title}</a>`).join('')}
    </ol>
  `)
}

/**
 * Renders the offline page
 */
export function offline() {
  return trim(
    `<p>Unfotunately this page is not available offline. Best go spin some vinyl...</p>
      <div>
          </div><img src="${offlineImg}" class="artwork" />
      </div>
      `
  )
}

/**
 * Renders the 404 page
 */
export function page404() {
  return trim(
    `<p>Oh noes, the page you're looking for doesn't exist! Try looking later...</p>
      <div>
          </div><img src="${fourOhFourImg}" class="artwork" />
      </div>
      `
  )
}

/**
 * Renders the 500 page
 */
export function page500() {
  return trim(
    `<p>Server hiccup. Nothing to see here...</p>
      <div>
          </div><img src="${fiveOhOhImg}" class="artwork" />
      </div>
      `
  )
}
