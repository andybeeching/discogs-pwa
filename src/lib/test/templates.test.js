import * as templates from '../templates'

import ROOT_STUB from '../../../__mocks__/stubs/root.json'
import NO_SEARCH_RESULT_STUB from '../../../__mocks__/stubs/searchNoResult.json'
import SEARCH_RESULT_STUB from '../../../__mocks__/stubs/searchResult.json'
import ARTIST_STUB from '../../../__mocks__/stubs/artist.json'
import RELEASE_STUB from '../../../__mocks__/stubs/release.json'
import RELEASE_LIST_STUB from '../../../__mocks__/stubs/releaseList.json'
import PAGINATION_STUB from '../../../__mocks__/stubs/pagination.json'
import MEMBERS_STUB from '../../../__mocks__/stubs/members.json'

describe('templates.js', () => {
  describe('root()', () => {
    it('should render "root" template', () => {
      const renderedTemplate = templates.root(ROOT_STUB)
      expect(renderedTemplate).toMatchSnapshot()
    })
  })

  describe('offline()', () => {
    it('should render offline template', () => {
      const renderedTemplate = templates.offline()
      expect(renderedTemplate).toMatchSnapshot()
    })
  })

  describe('page404()', () => {
    it('should render 404 page template', () => {
      const renderedTemplate = templates.page404()
      expect(renderedTemplate).toMatchSnapshot()
    })
  })

  describe('artistsSearchResults()', () => {
    const SEARCH_TERM = 'The Beatles'

    it('should render "no search results" template', () => {
      const renderedTemplate = templates.artistsSearchResults(
        NO_SEARCH_RESULT_STUB,
        SEARCH_TERM
      )
      expect(renderedTemplate).toMatchSnapshot()
    })

    it('should render "results found" template', () => {
      const renderedTemplate = templates.artistsSearchResults(
        SEARCH_RESULT_STUB,
        SEARCH_TERM
      )
      expect(renderedTemplate).toMatchSnapshot()
    })
  })

  describe('getPagination()', () => {
    const STUB_SLUG = '/endpoint?'

    it('should handle empty pagination object', () => {
      const renderedTemplate = templates.getPagination(
        PAGINATION_STUB['pagination-empty']
      )
      expect(renderedTemplate).toMatchSnapshot()
    })

    it('should handle first page pagination object', () => {
      const renderedTemplate = templates.getPagination(
        PAGINATION_STUB['pagination-first'],
        STUB_SLUG
      )
      expect(renderedTemplate).toMatchSnapshot()
    })

    it('should handle middle page pagination object', () => {
      const renderedTemplate = templates.getPagination(
        PAGINATION_STUB['pagination-middle'],
        STUB_SLUG
      )
      expect(renderedTemplate).toMatchSnapshot()
    })

    it('should handle last page pagination object', () => {
      const renderedTemplate = templates.getPagination(
        PAGINATION_STUB['pagination-last'],
        STUB_SLUG
      )
      expect(renderedTemplate).toMatchSnapshot()
    })
  })

  describe('artist()', () => {
    it('should render "artist" template', () => {
      const renderedTemplate = templates.artist(ARTIST_STUB)
      expect(renderedTemplate).toMatchSnapshot()
    })
  })

  describe('explodeArtists()', () => {
    it('should handle single artist', () => {
      const renderedTemplate = templates.explodeArtists(MEMBERS_STUB.one)
      expect(renderedTemplate).toMatchSnapshot()
    })
    it('should handle two artists', () => {
      const renderedTemplate = templates.explodeArtists(MEMBERS_STUB.two)
      expect(renderedTemplate).toMatchSnapshot()
    })
    it('should handle multiple artists', () => {
      const renderedTemplate = templates.explodeArtists(MEMBERS_STUB.multiple)
      expect(renderedTemplate).toMatchSnapshot()
    })
  })

  describe('releaseList()', () => {
    it('should render "release list" template', () => {
      const renderedTemplate = templates.releaseList(RELEASE_LIST_STUB)
      expect(renderedTemplate).toMatchSnapshot()
    })
  })

  describe('release()', () => {
    it('should render "release" template', () => {
      const renderedTemplate = templates.release(RELEASE_STUB)
      expect(renderedTemplate).toMatchSnapshot()
    })
  })
})
