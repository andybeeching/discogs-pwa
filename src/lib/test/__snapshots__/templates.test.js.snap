// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates.js artist() should render "artist" template 1`] = `
"
    <h2 class=\\"h2\\">Kings Of Convenience</h2>
    <p>Kings of Convenience are an indie folk-pop duo from Bergen, Norway. Consisting of Erlend Øye and Eirik Glambek Bøe, the musical group is known for their delicate tunes, calming voices, and intricately subtle guitar melodies. Both Øye and Bøe sing in their tracks, and both of them compose.</p>
    <p>Members: <a href=\\"/artist/82343?page=1\\">Erlend Øye</a>, and <a href=\\"/artist/462316?page=1\\">Eirik Glambek Bøe</a></p>
  "
`;

exports[`templates.js artistsSearchResults() should render "no search results" template 1`] = `
"<h2>Search results for \\"The Beatles\\"</h2>
    No search results found
    <ul class=\\"grid\\">
    </ul>"
`;

exports[`templates.js artistsSearchResults() should render "results found" template 1`] = `
"<h2>Search results for \\"The Beatles\\"</h2>
      <p>1 results were found, viewing 1 to 1</p>
    <ul class=\\"grid\\">
    <li class=\\"grid-item\\">
      <a href=\\"/artist/82730?page=1\\">
        <img src=\\"https://img.discogs.com/nSd8gtJIO-fTzvI6C5BrXxv4YuQ=/150x150/smart/filters:strip_icc():format(jpeg):mode_rgb():quality(40)/discogs-images/A-82730-1319532331.jpeg.jpg\\" />
        <span class=\\"desc\\">The Beatles</span>
      </a>
    </li>
    </ul>"
`;

exports[`templates.js explodeArtists() should handle multiple artists 1`] = `"<a href=\\"/artist/82343?page=1\\">Erlend Øye</a>, <a href=\\"/artist/462316?page=1\\">Eirik Glambek Bøe</a>, and <a href=\\"/artist/123456?page=1\\">Feist</a>"`;

exports[`templates.js explodeArtists() should handle single artist 1`] = `"<a href=\\"/artist/82343?page=1\\">Erlend Øye</a>"`;

exports[`templates.js explodeArtists() should handle two artists 1`] = `"<a href=\\"/artist/82343?page=1\\">Erlend Øye</a>, and <a href=\\"/artist/462316?page=1\\">Eirik Glambek Bøe</a>"`;

exports[`templates.js getPagination() should handle empty pagination object 1`] = `""`;

exports[`templates.js getPagination() should handle first page pagination object 1`] = `
"
    <ol class=\\"pagination\\">
      <li>
        <span class=\\"disabled-link\\">&laquo;</span>
      </li>
      <li>
        <span class=\\"disabled-link\\">Previous page</span>
      </li>
      <li>
        <a href=\\"/endpoint?page=2\\">Next page</a>
      </li>
      <li>
        <a href=\\"/endpoint?page=20\\">&raquo;</a>
      </li>
    </ol>
  "
`;

exports[`templates.js getPagination() should handle last page pagination object 1`] = `
"
    <ol class=\\"pagination\\">
      <li>
        <a href=\\"/endpoint?page=1\\">&laquo;</a>
      </li>
      <li>
        <a href=\\"/endpoint?page=19\\">Previous page</a>
      </li>
      <li>
        <span class=\\"disabled-link\\">Next page</span>
      </li>
      <li>
        <span class=\\"disabled-link\\">&raquo;</span>
      </li>
    </ol>
  "
`;

exports[`templates.js getPagination() should handle middle page pagination object 1`] = `
"
    <ol class=\\"pagination\\">
      <li>
        <a href=\\"/endpoint?page=1\\">&laquo;</a>
      </li>
      <li>
        <a href=\\"/endpoint?page=9\\">Previous page</a>
      </li>
      <li>
        <a href=\\"/endpoint?page=11\\">Next page</a>
      </li>
      <li>
        <a href=\\"/endpoint?page=20\\">&raquo;</a>
      </li>
    </ol>
  "
`;

exports[`templates.js release() should render "release" template 1`] = `
"<h2 class=\\"h2\\">Quiet Is The New Loud</h2>
    <p>By <a href=\\"/artist/12634?page=1\\">Kings Of Convenience</a></p>
    <div>
      <img src=\\"https://img.discogs.com/W1ebskmATPP9dg6a2nWlvKk7Tmw=/fit-in/600x586/filters:strip_icc():format(jpeg):mode_rgb():quality(90)/discogs-images/R-76745-1237046610.jpeg.jpg\\" class=\\"artwork\\" />
    </div>
    <h3>Tracklist</h3>
    <ol>
        <li>Winning A Battle, Losing The War</a><li>Toxic Girl</a><li>Singing Softly To Me</a><li>I Don't Know What I Can Save You From</a><li>Failure</a><li>The Weight Of My Words</a><li>The Girl From Back Then</a><li>Leaning Against The Wall</a><li>Little Kids</a><li>Summer On The Westhill</a><li>The Passenger</a><li>Parallel Lines</a>
    </ol>"
`;

exports[`templates.js releaseList() should render "release list" template 1`] = `
"<h3 class=\\"h3\\">Releases</h2>
    <ol class=\\"pagination\\">
      <li>
        <span class=\\"disabled-link\\">&laquo;</span>
      </li>
      <li>
        <span class=\\"disabled-link\\">Previous page</span>
      </li>
      <li>
        <a href=\\"/artist/undefined?page=2\\">Next page</a>
      </li>
      <li>
        <a href=\\"/artist/undefined?page=6\\">&raquo;</a>
      </li>
    </ol>
    <ul class=\\"grid\\">
    <li class=\\"grid-item\\">
      <a href=\\"/release/957026\\">
        <img src=\\"https://img.discogs.com/W3h5I5Xfo7m9t4924foueqrFAmA=/fit-in/150x150/filters:strip_icc():format(jpeg):mode_rgb():quality(40)/discogs-images/R-957026-1389622269-8515.jpeg.jpg\\" />
        <span class=\\"desc\\">Brave New World</span>
      </a>
    </li>
    </ul>
    <ol class=\\"pagination\\">
      <li>
        <span class=\\"disabled-link\\">&laquo;</span>
      </li>
      <li>
        <span class=\\"disabled-link\\">Previous page</span>
      </li>
      <li>
        <a href=\\"/artist/undefined?page=2\\">Next page</a>
      </li>
      <li>
        <a href=\\"/artist/undefined?page=6\\">&raquo;</a>
      </li>
    </ol>"
`;

exports[`templates.js root() should render "root" template 1`] = `
"<h2>Currently trending on Discogs</h2>
    <ul class=\\"grid\\">
    <li class=\\"grid-item\\">
      <a href=\\"/release/13486052\\">
        <img src=\\"https://img.discogs.com/a30BikrMMGCgDsFY20CFcafi6Cg=/fit-in/150x150/filters:strip_icc():format(jpeg):mode_rgb():quality(40)/discogs-images/R-13486052-1555118466-5371.jpeg.jpg\\" />
        <span class=\\"desc\\">The Mighty Boosh - The Mighty Boosh - The Complete Radio Series</span>
      </a>
    </li>
    </ul>"
`;
